third_party/move/move-compiler-v2/src/pipeline/ability_checker.rs

```diff
@@ -0,0 +1,259 @@
+        func_target.global_env().error(loc, err_msg)
+    }
+}
+
+/// Checks if the given temporary variable has constraint drop, and add diagnostics if not.
+fn check_drop_for_temp_with_msg(
+    func_target: &FunctionTarget,
+    t: TempIndex,
+    loc: &Loc,
+    err_msg: &str,
+) {
+    let ty = func_target.get_local_type(t);
+    check_drop(func_target, ty, loc, err_msg)
+}
+
+/// `t` is the local containing the reference being written to
+fn check_write_ref(target: &FunctionTarget, t: TempIndex, loc: &Loc) {
+    if let Type::Reference(_, ty) = target.get_local_type(t) {
+        check_drop(target, ty, loc, "write_ref: cannot drop")
+    } else {
+        panic!("ICE typing error")
+    }
+}
+
+fn check_key_for_struct(
+    target: &FunctionTarget,
+    mod_id: ModuleId,
+    struct_id: StructId,
+    insts: &[Type],
+    loc: &Loc,
+    err_msg: &str,
+) {
+    if !check_struct_inst(target, mod_id, struct_id, insts, &loc).has_ability(Ability::Key) {
+        target.global_env().error(loc, err_msg)
+    }
+}
+
+fn ty_param_abilities(ty_params: &[TypeParameter]) -> impl Fn(u16) -> AbilitySet + Copy + '_ {
+    |i| {
+        if let Some(tp) = ty_params.get(i as usize) {
+            tp.1.abilities
+        } else {
+            panic!("ICE unbound type parameter")
+        }
+    }
+}
+
+fn get_abilities<'a>(
+    target: &'a FunctionTarget,
+) -> impl Fn(ModuleId, StructId) -> (Vec<AbilitySet>, AbilitySet) + Copy + 'a {
+    |mid, sid| {
+        let qid = QualifiedId {
+            module_id: mid,
+            id: sid,
+        };
+        let struct_env = target.global_env().get_struct(qid);
+        let struct_abilities = struct_env.get_abilities();
+        let params_ability_constraints = struct_env
+            .get_type_parameters()
+            .iter()
+            .map(|tp| tp.1.abilities)
+            .collect_vec();
+        (params_ability_constraints, struct_abilities)
+    }
+}
+
+/// checks that the given type is instantiated with types satisfying their ability constraints
+/// on the type parameter
+fn check_struct_inst(
+    target: &FunctionTarget,
+    mid: ModuleId,
+    sid: StructId,
+    ty_args: &[Type],
+    loc: &Loc,
+) -> AbilitySet {
+    let ty_params = target.get_type_parameters();
+    ty::check_struct_inst(
+        mid,
+        sid,
+        ty_args,
+        ty_param_abilities(&ty_params),
+        get_abilities(target),
+        Some((loc, |loc: &Loc, msg: &str| target.global_env().error(loc, msg))),
+    )
+}
+
+fn check_fun_inst(target: &FunctionTarget, mid: ModuleId, fid: FunId, inst: &[Type], loc: &Loc) {
+    let qid = QualifiedId {
+        module_id: mid,
+        id: fid,
+    };
+    let fun_env = target.global_env().get_function(qid);
+    for (param, ty) in fun_env.get_type_parameters().iter().zip(inst.iter()) {
+        let required_abilities = param.1.abilities;
+        let given_abilities = check_instantiation(target, ty, loc);
+        // TODO: which field, why
```

Comment:
Puzzling comment.

Reply:
It's for the `target.global_env().error(loc, "invalid instantiation")`; we should include more information in the err messages.

Reply:
done